File: NET/Libraries/Treehopper.Libraries/Sensors/ProximitySensor.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;

namespace Treehopper.Libraries.Sensors
{
    public abstract class ProximitySensor : IProximity
    {
        public double Millimeters => Meters * 1000;
        public double Centimeters => Meters * 100;
        public double Inches => Meters * 39.3701;
        public double Feet => Meters * 3.28085;
        public abstract double Meters { get; }
        public bool AutoUpdateWhenPropertyRead { get; set; }
        public int AwaitPollingInterval { get; set; }

        public abstract Task Update();
    }
}

File: C++/API/src/libraries/Sensors/ProximitySensor.cpp
#pragma once

#include <Libraries/Sensors/ProximitySensor.h>

using namespace Treehopper::Libraries;

namespace Treehopper {
    namespace Libraries {
        namespace Sensors {
            double ProximitySensor::centimeters() {
                return meters() * 100;
            }

            double ProximitySensor::inches() {
                return meters() * 39.3701;
            }

            double ProximitySensor::millimeters() {
                return meters() * 1000;
            }

            double ProximitySensor::feet() {
                return meters() * 3.28085;
            }

            double ProximitySensor::meters() {
                if (autoUpdateWhenPropertyRead) update();

                return _meters;
            }
        }
    }
}
